### PR TITLE
[WebProfilerBundle] Use CPP

### PR DESCRIPTION
**Pull request by smnandre**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #... 
| License       | MIT

### CODE REVIEW

File: src/Symfony/Bundle/WebProfilerBundle/Controller/RouterController.php [hunk 1]
```diff
@@ -30,23 +30,19 @@
...
-        $this->routes = (null === $routes && $matcher instanceof RouterInterface) ? $matcher->getRouteCollection() : $routes;
-        $this->expressionLanguageProviders = $expressionLanguageProviders;
+    public function __construct(
+        private ?Profiler $profiler,
+        private Environment $twig,
+        private ?UrlMatcherInterface $matcher = null,
+        private ?RouteCollection $routes = null,
+        /**
+         * @var ExpressionFunctionProviderInterface[]
+         */
```
Comment (derrabus) [inline src/Symfony/Bundle/WebProfilerBundle/Controller/RouterController.php]: Let's turn this into a `@param` annotation on the constructor.

File: src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php [hunk 2]
```diff
@@ -26,14 +26,15 @@
 final class CodeExtension extends AbstractExtension
 {
     private string|FileLinkFormatter|array|false $fileLinkFormat;
-    private string $charset;
     private string $projectDir;
```
Comment (derrabus) [inline src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php]: Could be promoted as well.

File: src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php [hunk 3]
```diff
@@ -28,18 +28,16 @@
...
      */
     private $output;
 
     private int $stackLevel = 0;
 
-    public function __construct(?HtmlDumper $dumper = null)
-    {
-        $this->dumper = $dumper ?? new HtmlDumper();
+    public function __construct(
+        private ?HtmlDumper $dumper = new HtmlDumper(),
```
Comment (derrabus) [inline src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php]: Please revert. The property wasn't nullable before and you break constructor calls where `null` is passed explicitly.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/WebProfilerBundle/Controller/RouterController.php [hunk 4]
```diff
@@ -30,23 +30,19 @@
...
 class RouterController
 {
-    private ?Profiler $profiler;
-    private Environment $twig;
-    private ?UrlMatcherInterface $matcher;
-    private ?RouteCollection $routes;
-
     /**
-     * @var ExpressionFunctionProviderInterface[]
+     * @params ExpressionFunctionProviderInterface[] $expressionLanguageProviders
```
Comment (derrabus) [inline src/Symfony/Bundle/WebProfilerBundle/Controller/RouterController.php]:
```suggestion
     * @param ExpressionFunctionProviderInterface[] $expressionLanguageProviders
```

File: src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php [hunk 5]
```diff
@@ -25,15 +25,13 @@
...
 final class CodeExtension extends AbstractExtension
 {
-    private string|FileLinkFormatter|array|false $fileLinkFormat;
-    private string $charset;
-    private string $projectDir;
-
-    public function __construct(string|FileLinkFormatter $fileLinkFormat, string $projectDir, string $charset)
-    {
+    public function __construct(
+        private string|FileLinkFormatter|array|false $fileLinkFormat,
```
Comment (derrabus) [inline src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php]: I don't think we want to widen the constructor parameter type this way. Rule of thumb: if the parameter type and the property type don't match, we don't promote.

Comment (smnandre) [reply src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php]: Ok sorry, i revert

Comment (smnandre) [reply src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php]: ✅

Comment (smnandre) [reply src/Symfony/Bundle/WebProfilerBundle/Profiler/CodeExtension.php]: (thank you for the review 😄)

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @smnandre.

**Pull request merged**